### PR TITLE
Users/angel/fix samples fix login

### LINKED ISSUE #752: [BUG] Cannot login to Dynamics due issue in Login method
## Bug Report
Login method fails due trying to click to not existing controls, but do not check that is is exists before

**EasyRepro Version**
  - [ ] Microsoft Dynamics 365 Online Version 8.2 (8.2.x) (DB 8.2.x) online
  - [ ] Microsoft Dynamics 365 Online Version 9.0 (9.0.x) (DB 9.0.x) online
  - [ ] Microsoft Dynamics 365 Online Version 9.0.2 (9.0.2) (DB 9.0.2) online
  - [ + ] Microsoft Dynamics 365 Online Version 9.1 (9.1.x) (DB 9.1.x) online
  
**UCI or Classic Web**
  - [ + ] UCI (XrmApp)
  - [ ] Classic Web (XrmBrowser)

**Online or On Premise**
  - [ + ] Online
  - [ ] On Premise

**Browser**
  - [ + ] Chrome
  - [ ] Firefox
  - [ ] IE
  - [ ] Edge

**Describe the bug**  
During login test trying to click to not existing controls, but do not check that is is exists before

```csharp
driver.ClickWhenAvailable(By.Id("aadTile"));
driver.ClickWhenAvailable(By.Id("use_another_account_link"));
```
Stack Trace:
```
Test method Microsoft.Dynamics365.UIAutomation.Sample.UCI.OpenContactUCI.UCITestOpenContactRetrieveHeaderValues threw exception: 
System.InvalidOperationException: Unable to click element.
    at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.<>c__DisplayClass50_0.<WaitUntilClickable>b__0() in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 500
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntil(ISearchContext driver, Func`2 searchFunc, Nullable`1 timeout, Action`1 successCallback, Action failureCallback) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 558
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntilClickable(ISearchContext driver, By by, Nullable`1 timeout, Action`1 successCallback, Action failureCallback) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 508
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntilClickable(ISearchContext driver, By by, Nullable`1 timeout, Action`1 successCallback, String exceptionMessage) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 500
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.ClickWhenAvailable(ISearchContext driver, By by, Nullable`1 timeout, String errorMessage) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 38
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.WebClient.Login(IWebDriver driver, Uri uri, SecureString username, SecureString password, SecureString mfaSecrectKey, Action`1 redirectAction) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\WebClient.cs:line 127
   at Microsoft.Dynamics365.UIAutomation.Browser.DelegateBrowserCommand`6.ExecuteCommand(IWebDriver driver, Object[] params) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\DelegateBrowserCommand.cs:line 163
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserCommand`1.Execute[T1,T2,T3,T4,T5,T6,T7,T8,T9](IWebDriver driver, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5, T6 p6, T7 p7, T8 p8, T9 p9) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserCommand.cs:line 136
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserCommand`1.Execute[T1,T2,T3,T4,T5](IWebDriver driver, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserCommand.cs:line 66
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserPage.Execute[TResult,T1,T2,T3,T4,T5](BrowserCommandOptions options, Func`7 delegate, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserPage.cs:line 242
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.WebClient.Login(Uri orgUri, SecureString username, SecureString password, SecureString mfaSecrectKey, Action`1 redirectAction) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\WebClient.cs:line 116
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.OnlineLogin.Login(Uri orgUrl, SecureString username, SecureString password, SecureString mfaSecrectKey) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\Elements\OnlineLogin.cs:line 38
   at Microsoft.Dynamics365.UIAutomation.Sample.UCI.OpenContactUCI.UCITestOpenContactRetrieveHeaderValues() in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Sample\UCI\Read\OpenContact.cs:line 150
```

**Code to reproduce**  
Run UCITestOpenContactRetrieveHeaderValues test

```csharp
xrmApp.OnlineLogin.Login(_xrmUri, _username, _password);
```
 
**Expected behavior**  
It should be verification that controls exists before click
```csharp
if (driver.IsVisible(By.Id("aadTile")))
	driver.ClickWhenAvailable(By.Id("aadTile"));
if (driver.IsVisible(By.Id("use_another_account_link")))
	driver.ClickWhenAvailable(By.Id("use_another_account_link"
```));

### PR DESCRIPTION
**Pull request by AngelRodriguez8008**
### Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (updates to documentation, formatting, etc.)

### Description
Fix #752 (splited branch from #754)

### Issues addressed
Login is performed now but after Login method is hang on EnterOneTimeCode(driver, mfaSecrectKey); command inside Login method. [see more](https://github.com/microsoft/EasyRepro/issues/752#issuecomment-583972275)

### All submissions:

- [x] My code follows the code style of this project.
- [x] Do existing samples that are effected by this change still run?
- [x] I have added samples for new functionality. 
- [x] I raise detailed error messages when possible.
- [x] My code does not rely on labels that have the option to be hidden.

### Which browsers was this tested on?

- [x] Chrome
- [ ] Firefox
- [ ] IE
- [ ] Edge

### CODE REVIEW

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 1]
```diff
@@ -901,56 +913,47 @@ internal BrowserCommandResult<bool> ConfirmationDialog(bool ClickConfirmButton)
             });
         }
 
-        internal BrowserCommandResult<bool> AssignDialog(Dialogs.AssignTo to, string userOrTeamName)
+        internal BrowserCommandResult<bool> AssignDialog(Dialogs.AssignTo to, string userOrTeamName = null)
         {
+            userOrTeamName = userOrTeamName?.Trim() ?? string.Empty;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:918]: This Fix to Assign was unintentional add to this PR but is not too big

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/Elements/Dialog.cs [hunk 2]
```diff
@@ -58,7 +58,7 @@ public void CloseOpportunity(double revenue, DateTime closeDate, string descript
         /// </summary>
         /// <param name="to">Enum used to assign record to user or team</param>
         /// <param name="userOrTeamName">Name of the user or team to assign to</param>
-        public void Assign(Dialogs.AssignTo to, string userOrTeamName = "")
+        public void Assign(AssignTo to, string userOrTeamName = null)
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/Elements/Dialog.cs:61]: This Fix to Assign was unintentional add to this PR but is not too big

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 3]
```diff
@@ -66,37 +67,23 @@ internal BrowserCommandResult<bool> InitializeModes()
...
-        internal void WaitForMainPage()
-        {
-            IWebDriver driver = this.Browser.Driver;
-            driver.WaitUntilVisible(By.XPath(Elements.Xpath[Reference.Login.CrmMainPage]));
-            driver.WaitForPageToLoad();
-            if (driver.HasElement(By.XPath(Elements.Xpath[Reference.Login.CrmUCIMainPage])))
-            {
-                driver.WaitForTransaction();
-            }
+        internal bool WaitForMainPage(TimeSpan timeout, string errorMessage)
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:70]: This method was refactored,  to be more flexible, is used in Login logic to wait

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 4]
```diff
@@ -213,41 +202,39 @@ private static void EnterPassword(IWebDriver driver, SecureString password)
...
+                var oneTimeCode = GenerateOneTimeCode(mfaSecrectKey);
+                SetInputValue(driver, input, oneTimeCode, 1.Seconds());
+                input.Submit();
+                return true; // input found & code was entered
+            }
+            catch (Exception e)
+            {
+                var message = $"An Error occur entering OTC. Exception: {e.Message}";
+                Trace.TraceInformation(message);
+                throw driver.Throw<InvalidOperationException>(message, e);
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs]: the Loop logic was moved to the caller context

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 5]
```diff
@@ -213,41 +202,39 @@ private static void EnterPassword(IWebDriver driver, SecureString password)
...
             => driver.WaitUntilAvailable(By.XPath(Elements.Xpath[Reference.Login.OneTimeCode]), TimeSpan.FromSeconds(2));
 
-        private static void ClickStaySignedIn(IWebDriver driver)
+        private static bool ClickStaySignedIn(IWebDriver driver)
         {
             var xpath = By.XPath(Elements.Xpath[Reference.Login.StaySignedIn]);
-            driver.WaitUntilVisible(xpath, new TimeSpan(0, 0, 5),
-                e => driver.ClickWhenAvailable(xpath));
+            var element = driver.ClickIfVisible(xpath, 5.Seconds());
+            return element != null;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:237]: Better implementation

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 6]
```diff
@@ -271,7 +258,22 @@ internal BrowserCommandResult<LoginResult> PassThroughLogin(Uri uri)
...
+                        {
+                            driver.WaitForPageToLoad();
+                            driver.WaitForTransaction();
+                        }
+                        else
+                            //else we landed on the Web Client main page or app picker page
+                            SwitchToDefaultContent(driver);
+                    },
+                    () => driver.Throw("Load Main Page Fail.")
+                );
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:276]: WaitForLoginPage is just a variation of WaitForMainPage with more time & throw an Error if this is not shown.

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 7]
```diff
@@ -337,12 +339,14 @@ internal BrowserCommandResult<bool> OpenApp(string appName, int thinkTime = Cons
...
                 driver.SwitchTo().DefaultContent();
 
                 //Handle left hand Nav in Web Client
-                var success = TryOpenAppFromMenu(driver, appName, AppReference.Navigation.WebAppMenuButton) ||
-                              TryOpenAppFromMenu(driver, appName, AppReference.Navigation.UCIAppMenuButton) ||
-                              TryToClickInAppTile(appName, driver);
+                var success = TryToClickInAppTile(appName, driver) ||
+                              TryOpenAppFromMenu(driver, appName, AppReference.Navigation.WebAppMenuButton) ||
+                              TryOpenAppFromMenu(driver, appName, AppReference.Navigation.UCIAppMenuButton);
+
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:349]: try to clic first on AppTile it is faster

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 8]
```diff
@@ -355,52 +359,60 @@ internal BrowserCommandResult<bool> OpenApp(string appName, int thinkTime = Cons
...
-                throw new InvalidOperationException($"App Button {appMenuButton} not found.", e);
-            }
+            bool found = false;
+            var xpathToAppMenu = By.XPath(AppElements.Xpath[appMenuButton]);
+            driver.WaitUntilClickable(xpathToAppMenu, TimeSpan.FromSeconds(5),
+                        appMenu =>
+                        {
+                            appMenu.Click(true);
+                            OpenAppFromMenu(driver, appName);
+                            found = true;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:369]: Wait 5 seconds for load, if required

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 9]
```diff
@@ -1903,6 +1906,19 @@ internal BrowserCommandResult<bool> SetValue(string field, string value)
...
+            input.SendKeys(Keys.Backspace);
+            driver.WaitForTransaction();
+
+            if (string.IsNullOrWhiteSpace(value))
+                return;
+
+            input.SendKeys(value, true);
+            driver.WaitForTransaction();
+            ThinkTime(thinktime ?? 3.Seconds());
+        }
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:1921]: this method is used to enter the One Time Code, but will be also used in future implementations.

File: Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs [hunk 10]
```diff
@@ -598,6 +618,18 @@ public static IWebElement WaitUntilClickable(this ISearchContext driver, By by,
...
+            driver.Quit();
+            return new InvalidOperationException(message, innerException);
+        }
+
+        public static Exception Throw<T>(this IWebDriver driver, string message, Exception innerException = null) where T : Exception
+        {
+            driver.Quit();
+            return (T) Activator.CreateInstance(typeof(T), message, innerException);
+        }
+
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs]: An Exception in Login Methods do left the Browser open (strange but true), calling this methods to throw the Exception solve this problem. See my comment [hier](https://github.com/microsoft/EasyRepro/issues/752#issuecomment-584012093)

File: Microsoft.Dynamics365.UIAutomation.Browser/Extensions/TimeExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+using System;
+
+namespace Microsoft.Dynamics365.UIAutomation.Browser
+{
+    public static class TimeExtensions
+    {
+        public static TimeSpan Seconds(this int value)
+            => TimeSpan.FromSeconds(value);
+    }
+}
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Browser/Extensions/TimeExtensions.cs:13]: sintaxis sugar: make the code more expressive. instead new TimeSpan(0, 0, 60) or TimeSpan.FromSeconds(60) => 60.Seconds(), => ex: Wait(60.Seconds())  or ThinkTime(60.Seconds())

**Review by AngelRodriguez8008 — COMMENTED** [review]
Hi @dtu11, I hope this is fine now. Thanks

File: Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs [hunk 12]
```diff
@@ -22,8 +22,16 @@ public class TestsBase
...
+            try
+            {  
+                CreateApp();
+                NavigateToHomePage();
+            }
+            catch
+            {
+                CloseApp();
+                throw;
+            }
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs:34]: If Login or Navigate Fail close the Browser & rethrow the exception

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 13]
```diff
@@ -165,21 +152,23 @@ private LoginResult Login(IWebDriver driver, Uri uri, SecureString username, Sec
...
+            do
+            {
+                entered = EnterOneTimeCode(driver, mfaSecrectKey);
+                success = ClickStaySignedIn(driver) || IsUserAlreadyLogged();
+                attempts++;
+            }
+            while (!success && attempts <= Constants.DefaultRetryAttempts); // retry to enter the otc-code, if its fail & it is requested again 
+           
+            if (entered && !success)
+                throw new InvalidOperationException("Somethig got wrong entering the OTC. Please check the MFA-SecrectKey in configuration.");
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:166]: retry to enter the otc-code, if its fail & it is requested again, until stay signeg is shown or the main page is loaded

File: Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs [hunk 14]
```diff
@@ -213,41 +202,39 @@ private static void EnterPassword(IWebDriver driver, SecureString password)
...
+                var oneTimeCode = GenerateOneTimeCode(mfaSecrectKey);
+                SetInputValue(driver, input, oneTimeCode, 1.Seconds());
+                input.Submit();
+                return true; // input found & code was entered
+            }
+            catch (Exception e)
+            {
+                var message = $"An Error occur entering OTC. Exception: {e.Message}";
+                Trace.TraceInformation(message);
+                throw new InvalidOperationException(message, e);
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Api.UCI/WebClient.cs:225]: Loop was moved to the call context

**Review by AngelRodriguez8008 — COMMENTED** [review]
now smaller

**Review by TYLEROL — APPROVED** [review]

**Pull request merged**